### DISCUSSION TITLE
How to override component-level CSS variables?

### DISCUSSION THREAD

**Discussion by martinbean** [twbs/bootstrap#36369] [Q&A] [answered]
I’ve just installed the 5.2 beta as excited to try out CSS custom properties (variables)! I seem to be misunderstanding how they work, though.

In cards, I can see there are properties like `--bs-box-shadow`. So, I added a `<style>` tag with the following rule:

```html
<style>
body {
    --bs-card-box-shadow: 0 5px 10px black;
}
</style>
```

This doesn’t seem to have any effect, though. If I inspect a card on my page, it doesn’t seem to be picking up my defined value and instead has a “blank” value:

![Screenshot 2022-05-16 at 15 53 46](https://user-images.githubusercontent.com/167312/168621764-b94fdbd1-97ed-48e0-a36b-a8f683cc0772.png)

![Screenshot 2022-05-16 at 15 53 34](https://user-images.githubusercontent.com/167312/168621613-b289ee5b-7a6b-4568-8221-46f317cfc370.png)

It _does_ work if I define the value in a `.card` selector:

```html
<style>
.card {
    --bs-card-box-shadow: 0 5px 10px black;
}
</style>
```

But this to me defeats the point of having variables if we then have to define component-specific variables in many selectors?

I’m asking as I’m trying to introduce theming to an application and I’d like to just be able to set a load of custom properties on the `body` element (overriding Bootstrap’s set in `:root`), but this isn’t going to be feasible if I have to define property values in various CSS class selectors.

Why do property values set in a selector like `body` not propagate down to the components? Why do I have to define property values in component-specific selectors like `.card`?

**Comment by mdo** [comment, marked as answer]
Our implementation currently puts component-specific variables on the component level as to not flood the `:root` element. For example, if you're customizing your build by not importing half our components, we don't want those to still be there or require additional enabling/disabling on a per-component basis.

You will notice though in the `_root.scss` stylesheet that we have a growing number of global CSS variables that we reuse across the project for shared styles. Things like colors, fonts, borders, and more. Those are variables that can be overridden on the `:root` or `<body>` and have their changes globally inherited. This set of root variables will continue to grow with dark mode support coming in #35857 where we add additional, more nuanced shared colors (kind of like Apple's HIG).

With v5 at least, one of the main goals of adding CSS variables is to enable color mode support. But, it's not a feasible solution to have to customize every single component's colors to make that happen. As such, we've opted for fewer global variables and more component variables. This isn't a perfect setup yet—mixing Sass functionality with CSS variables is a little messy. However, this will continue to be iterated on and refined heading into a v6 eventually.

**Reply by martinbean** [reply]
Thanks for the explanation, @mdo. Not flooding `:root` with component-specific variables if those components aren’t being compiled into the final stylesheet makes sense.

Love the work that’s been done on this front so far, though. And keen to see where it heads in the future. Kudos to every one who’s been involved!

**Reply by jez9999** [reply]
@mdo I've just come across this same issue.  Why doesn't this work from a CSS perspective, though?  If I define `:root { --bs-navbar-color: blue !important; }`, my variable is overridden by the `--bs-navbar-color` definition in `_navbar.scss`.  Why?  My variable has `!important` on it.  My reading of the CSS variables documentation is that `!important` should make my variable override the Bootstrap one anyway.

**Reply by jez9999** [reply]
OK, I was forgetting how CSS worked.  This does indeed override Bootstrap component-level variables: `:root * { --bs-navbar-color: blue !important; }` - the question is, is that considered bad practice?  It might be for normal styling but for CSS variables it doesn't seem like such a bad idea.
